devito/ir/iet/scheduler.py

```diff
@@ -0,0 +1,158 @@
+from collections import OrderedDict
+
+import numpy as np
+
+from devito.cgen_utils import Allocator
+from devito.dimension import LoweredDimension
+from devito.ir.iet import (Expression, LocalExpression, Element, Iteration, List,
+                           UnboundedIndex, MetaCall, MapExpressions, Transformer,
+                           NestedTransformer, SubstituteExpression, iet_analyze,
+                           compose_nodes, filter_iterations, retrieve_iteration_tree)
+from devito.tools import filter_ordered, flatten
+from devito.types import Scalar
+
+__all__ = ['iet_build', 'iet_insert_C_decls']
+
+
+def iet_build(clusters, dtype):
+    """
+    Create an Iteration/Expression tree (IET) given an iterable of :class:`Cluster`s.
+    The nodes in the returned IET are decorated with properties deriving from
+    data dependence analysis.
+    """
+    # Clusters -> Iteration/Expression tree
+    iet = iet_make(clusters, dtype)
+
+    # Data dependency analysis. Properties are attached directly to nodes
+    iet = iet_analyze(iet)
+
+    # Substitute derived dimensions (e.g., t -> t0, t + 1 -> t1)
+    # This is postponed up to this point to ease /iet_analyze/'s life
+    subs = {}
+    for tree in retrieve_iteration_tree(iet):
+        uindices = flatten(i.uindices for i in tree)
+        subs.update({i.expr: LoweredDimension(name=i.index.name, origin=i.expr)
+                     for i in uindices})
```

Comment:
An aside comment that needs not be addressed in this PR, but has bugged me in another context: The role of `LoweredDimension` and `UnboundIndex` have to be re-though. The problem is that they effectively express the same thing at two different levels (pure expressions and IET), but will not symbolically match each other, making it awkward to find corresponding pairs. Food for thought...

Reply:
Agreed. In fact, I tried to kill off ``LoweredDimension`` in this PR, but I failed. I think I'm getting close though